### DISCUSSION TITLE
Vercel + serverDependenciesToBundle + ESM

### DISCUSSION THREAD

**Discussion by raulfdm** [remix-run/remix#4917] [Q&A]
Hey folks, I hope you're right.

I encountered the following, and I'd love to hear opinions on that.

I started a Remix project (no stack) and chose to deploy on vercel inside a workspace controlled by `pnpm`.

In this project, I use a dependency called `query-string` to handle some parameters.

This dependency, however, is full ESM, which means I need to include it on my `serverDependenciesToBundle`.

While running it locally, it complained to me because one of the `query-string` dependencies also ESM (`filter-obj`). Then I went again to `serverDependenciesToBundle` and added this other ESM lib.

Locally (running `@remix/server`), everything works just fine, however, when I ship the code to Vercel, the function throws an error saying:

```
2022-12-19T21:04:23.932Z	undefined	ERROR	Cannot find module 'decode-uri-component'
Require stack:
- /var/task/apps/website-remix/api/index.js
2022-12-19T21:04:23.932Z	undefined	ERROR	Did you forget to add it to "dependencies" in `package.json`?
RequestId: 92304a19-f246-4d51-bbc1-e03da570968f Error: Runtime exited with error: exit status 1
Runtime.ExitError
```

After some debugging, I noticed that this is because when I added `query-string` to be bundled, it was included in the output files (`root/api/index.js`), but its dependencies (listed on its package.json) were just being "required":

```js
var import_decode_uri_component = __toESM(require("decode-uri-component"), 1)
```

My guess is because this is a dependency of a dependency (query-string), when it tries to require, and this is not listed on my package.json, node cannot find it, especially because I'm using `pnpm` and it doesn't flat the `node_modules`.

My workaround was adding all dependencies of `query-string` to be bundled up, and now everything is working again, but I was wondering if there's a better (scalable) way of doing this.

For comparison, I have the exact same project in NextJS, and I don't have to either include this dependency to be bundled and everything works just fine, probably because of the way the bundle is generated.

**Comment by arashi-dev** [comment]
any solution?

```
2024-01-24T21:56:15.820Z	undefined	ERROR	Cannot find module 'next/dist/server/next-server.js'
Require stack:
- /var/task/___next_launcher.cjs
2024-01-24T21:56:15.821Z	undefined	ERROR	Did you forget to add it to "dependencies" in `package.json`?
```
